torch/csrc/api/src/optim/optimizer.cpp

```diff
@@ -11,10 +10,92 @@
+  return params_;
+}
+
+const std::vector<Tensor>& OptimizerParamGroup::params() const {
+  return params_;
+}
+
+std::unique_ptr<OptimizerParamState> OptimizerParamState::clone() const {
+  TORCH_CHECK(false,
+      "clone() has not been implemented for torch::optim::OptimizerParamState. ",
+      "Subclass torch::optim::OptimizerCloneableParamState<YourOptimizerParamState> ",
+      "instead of torch::optim::OptimizerParamState to inherit the ability to clone.");
+}
+
+void OptimizerParamState::serialize(torch::serialize::InputArchive& archive) {
+  TORCH_CHECK(false,
+    "void serialize(torch::serialize::InputArchive& archive) has not been implemented for torch::optim::OptimizerParamState. ",
+    "You must override it in your subclass of torch::optim::OptimizerCloneableParamState<YourOptimizerParamState>.");
+}
+
+void OptimizerParamState::serialize(torch::serialize::OutputArchive& archive) const {
+  TORCH_CHECK(false,
+    "void serialize(torch::serialize::OutputArchive& archive) has not been implemented for torch::optim::OptimizerParamState. ",
+    "You must override it in your subclass of torch::optim::OptimizerCloneableParamState<YourOptimizerParamState>.");
+}
+
+std::unique_ptr<OptimizerOptions> OptimizerOptions::clone() const {
+  TORCH_CHECK(false,
+      "clone() has not been implemented for torch::optim::OptimizerOptions. ",
+      "Subclass torch::optim::OptimizerCloneableOptions<YourOptimizerOptions> ",
+      "instead of torch::optim::OptimizerOptions to inherit the ability to clone.");
+}
+
+void OptimizerOptions::serialize(torch::serialize::InputArchive& archive) {
+  TORCH_CHECK(false,
+    "void serialize(torch::serialize::InputArchive& archive) has not been implemented for torch::optim::OptimizerOptions. ",
+    "You must override it in your subclass of torch::optim::OptimizerCloneableOptions<YourOptimizerOptions>.");
+}
+
+void OptimizerOptions::serialize(torch::serialize::OutputArchive& archive) const {
+  TORCH_CHECK(false,
+    "void serialize(torch::serialize::OutputArchive& archive) has not been implemented for torch::optim::OptimizerOptions. ",
+    "You must override it in your subclass of torch::optim::OptimizerCloneableOptions<YourOptimizerOptions>.");
+}
+
 namespace detail {
 OptimizerBase::OptimizerBase(std::vector<Tensor> parameters)
     : parameters_(std::move(parameters)) {}
 
+void OptimizerBase::add_param_group(const OptimizerParamGroup& param_group) {
+  for (const auto& param : param_group.params()) {
+    TORCH_CHECK(param.is_leaf(), "can't optimize a non-leaf Tensor");
+  }
+  OptimizerParamGroup param_group_(param_group.params());
+  if (!param_group.has_options()) {
+    param_group_.set_options(defaults_->clone());
+  } else {
+    param_group_.set_options(param_group.options().clone());
+  }
+  for (const auto& p : param_group_.params()) {
+    TORCH_CHECK(state_.count(c10::guts::to_string(p.unsafeGetTensorImpl())) == 0,
+      "some parameters appear in more than one parameter group");
```

Comment:
It would be awesome to add a test for this runtime error :D

Reply:
Hey! I think we should not have a function returning a reference to param_group_ because then we might get unwanted behavior like a user can add a param_group with params existing in other param_groups by:
```
auto& params_groups = optimizer.param_groups();
params_groups.push_back(OptimizerParamGroup(params));```
and this behavior should not be allowed

Reply:
Yes I agreed and we should remove the function returning non-const reference to param_group_.

Reply:
I do think we still need to add a test for this runtime error though, as we could still have param groups that share tensors.